compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3479,18 +3480,24 @@ BType getTypeForFiniteTypeValuesAssignableToType(BFiniteType finiteType, BType t
      * Method to retrieve a type representing all the member types of a union type that are assignable to
      * the target type.
      *
-     * @param intersectionContext
      * @param unionType  the union type
      * @param targetType the target type
+     * @param intersectionContext
+     * @param visitedTypes cache to capture visited types
      * @return           a single type or a new union type if at least one member type of the union type is
      *                      assignable to targetType, else semanticError
      */
     BType getTypeForUnionTypeMembersAssignableToType(BUnionType unionType, BType targetType, SymbolEnv env,
-                                                     IntersectionContext intersectionContext) {
+                                                     IntersectionContext intersectionContext,
+                                                     LinkedHashSet<BType> visitedTypes) {
         List<BType> intersection = new LinkedList<>();
+        if (!visitedTypes.add(unionType)) {
```

Comment:
Don't we need the type pair here? Is it always enough to add just the union type?